src/main/java/com/commercetools/sync/states/StateSync.java

```diff
@@ -0,0 +1,445 @@
+        @Nonnull final Set<String> missingTransitionParentStateKeys) {
+
+        missingTransitionParentStateKeys.forEach(missingParentKey ->
+            statistics.addMissingDependency(missingParentKey, newState.getKey()));
+
+        return unresolvedTransitionsService.save(
+            new WaitingToBeResolvedTransitions(newState, missingTransitionParentStateKeys));
+    }
+
+    @Nonnull
+    private CompletionStage<Void> syncDraft(
+        @Nonnull final Set<State> oldStates,
+        @Nonnull final StateDraft newStateDraft) {
+
+        final Map<String, State> oldStateMap =
+            oldStates.stream().collect(toMap(State::getKey, identity()));
+
+        return stateReferenceResolver
+            .resolveReferences(newStateDraft)
+            .thenCompose(resolvedDraft -> {
+
+                final State oldState = oldStateMap.get(newStateDraft.getKey());
+
+                return ofNullable(oldState)
+                    .map(state -> buildActionsAndUpdate(oldState, resolvedDraft))
+                    .orElseGet(() -> applyCallbackAndCreate(resolvedDraft));
+
+            })
+            .exceptionally(completionException -> {
+                final String errorMessage = format(FAILED_TO_PROCESS, newStateDraft.getKey(),
+                    completionException.getMessage());
+                handleError(errorMessage, completionException, 1);
+                return null;
+            });
+    }
+
+    /**
+     * Given a state draft, this method applies the beforeCreateCallback and then issues a create request to the
+     * CTP project to create the corresponding State.
+     *
+     * @param stateDraft the state draft to create the state from.
+     * @return a {@link CompletionStage} which contains an empty result after execution of the create.
+     */
+    @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION") // https://github.com/findbugsproject/findbugs/issues/79
+    @Nonnull
+    private CompletionStage<Void> applyCallbackAndCreate(@Nonnull final StateDraft stateDraft) {
+        return syncOptions
+            .applyBeforeCreateCallBack(stateDraft)
+            .map(draft -> stateService
+                .createState(draft)
+                .thenAccept(stateOptional -> {
+                    if (stateOptional.isPresent()) {
+                        readyToResolve.add(stateDraft.getKey());
+                        statistics.incrementCreated();
+                    } else {
+                        statistics.incrementFailed();
+                    }
+                })
+            )
+            .orElse(completedFuture(null));
+    }
+
+    @Nonnull
+    @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION")
+    private CompletionStage<Void> buildActionsAndUpdate(
+        @Nonnull final State oldState,
+        @Nonnull final StateDraft newState) {
+
+        final List<UpdateAction<State>> updateActions = buildActions(oldState, newState);
+        buildSetTransitionsAction(oldState, newState).ifPresent(updateActions::add);
```

Comment:
@ahalberkamp it looks like to me that you break the feature in here to pass your new tests 😢 , the draft will have a key so you can not create with keytoId cache, because in normal usage draft's transition id will be a key, and you can not create a reference with a key. https://github.com/commercetools/commercetools-sync-java/commit/353be7375a2da6ef9b46bfac6ed86820a103db63

Reply:
No, I have fixed the logic, because before updating the transitions, the "key-based" transitions get "resolved" to the id`s of the target system.

Please check: https://github.com/commercetools/commercetools-sync-java/blob/rebased-409/src/main/java/com/commercetools/sync/states/StateSync.java#L231

After his, the transition  already has the proper id of the target system, and so the ID don`t have to be mapped using the "keytoId" map.

Reply:
With which commit that you fix something, I could find any changes.

Reply:
Also in here, buildSetTransitionsAction should be part of the build actions, before we were passing the map but according to your change, it's not needed anymore. 

https://github.com/commercetools/commercetools-sync-java/blob/353be7375a2da6ef9b46bfac6ed86820a103db63/src/main/java/com/commercetools/sync/states/StateSync.java#L282-L284

Reply:
I removed the key-mapping here: https://github.com/commercetools/commercetools-sync-java/commit/353be7375a2da6ef9b46bfac6ed86820a103db63#diff-af209f84ba847bda9fce3f29e5605c54R174 ,which fixed the described problems